src/DataStructures/Tensor/EagerMath/DotProduct.hpp

```diff
@@ -20,16 +21,27 @@
  * or \f$A_a B_b \delta^{ab}\f$ for input one forms \f$A_a\f$ and \f$B_b\f$.
  */
 template <typename DataType, typename Index>
-Scalar<DataType> dot_product(
+void dot_product(
+    const gsl::not_null<Scalar<DataType>*> dot_product,
     const Tensor<DataType, Symmetry<1>, index_list<Index>>& vector_a,
     const Tensor<DataType, Symmetry<1>, index_list<Index>>& vector_b) noexcept {
-  auto dot_product = make_with_value<Scalar<DataType>>(vector_a, 0.);
-  for (size_t d = 0; d < Index::dim; ++d) {
-    get(dot_product) += vector_a.get(d) * vector_b.get(d);
+  get(*dot_product) = vector_a.get(0) * vector_b.get(0);
```

Comment:
This wasn't factored out for performance reasons, it was factored out because if we are using `not_null` we cannot rely on the incoming data being zero, thus we must assign to it. Doing the compile-time loop unrolling doesn't always help. I've explored this in the past and it ends up being hardware and other instructions around that line dependent. At this point for GRMHD the only real way we're going to see any speed up is by coming up with a different formulation or wicked fast inversion schemes. If you want you can see what differences might be, but I would advise against it because I don't think it's a valuable way of spending time right now. The main bottleneck is almost always memory allocations. Once those are squashed then at least for GRMHD we are waiting on inversion, after which it is computing derivatives.

Reply:
That makes sense; thanks, it helps to understand where the most expensive bits are (at least at current). Do you have a separate branch with your benchmark functions? I'd find it useful for when I want to look at performance questions in the future to see the techniques you find most reliable.

Reply:
Good question! I wrote a thing: https://github.com/nilsdeppe/spectre/wiki/Benchmark-Codes and I'll update it with more benchmarks as I do them. Hopefully that helps!

Reply:
Thanks! that's very helpful. :+1: